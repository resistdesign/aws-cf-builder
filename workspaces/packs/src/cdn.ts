import { createResourcePack } from '@aws-cf-builder/utils';

export type AddCDNConfig = {
  id: string;
  hostedZoneId: any;
  domainName: any;
  certificateArn: any;
  fileStorageId: string;
};

export const addCDN = createResourcePack(({ id, hostedZoneId, domainName, certificateArn, fileStorageId }: AddCDNConfig) => {
  return {
    Resources: {
      [`${id}CloudFront`]: {
        Type: 'AWS::CloudFront::Distribution',
        DependsOn: fileStorageId,
        Properties: {
          DistributionConfig: {
            Aliases: [domainName],
            ViewerCertificate: {
              AcmCertificateArn: certificateArn,
              SslSupportMethod: 'sni-only',
              MinimumProtocolVersion: 'TLSv1.1_2016',
            },
            DefaultCacheBehavior: {
              ForwardedValues: {
                QueryString: false,
              },
              TargetOriginId: {
                'Fn::Sub': [
                  'S3-${S3BucketName}',
                  {
                    S3BucketName: domainName,
                  },
                ],
              },
              ViewerProtocolPolicy: 'redirect-to-https',
            },
            DefaultRootObject: 'index.html',
            Enabled: true,
            IPV6Enabled: false,
            HttpVersion: 'http2',
            Origins: [
              {
                DomainName: {
                  'Fn::Sub': [
                    '${S3BucketName}.s3.amazonaws.com',
                    {
                      S3BucketName: domainName,
                    },
                  ],
                },
                Id: {
                  'Fn::Sub': [
                    'S3-${S3BucketName}',
                    {
                      S3BucketName: domainName,
                    },
                  ],
                },
                S3OriginConfig: {
                  OriginAccessIdentity: '',
                },
              },
            ],
            CustomErrorResponses: [
              {
                ErrorCachingMinTTL: 300,
                ErrorCode: 404,
                ResponseCode: 200,
                ResponsePagePath: '/index.html',
              },
              {
                ErrorCachingMinTTL: 300,
                ErrorCode: 403,
                ResponseCode: 200,
                ResponsePagePath: '/index.html',
              },
            ],
            PriceClass: 'PriceClass_All',
          },
        },
      },
      [`${id}Route53Record`]: {
        Type: 'AWS::Route53::RecordSet',
        DependsOn: [`${id}CloudFront`],
        Properties: {
          HostedZoneId: hostedZoneId,
          Type: 'A',
          Name: {
            'Fn::Sub': [
              '${DomainName}.',
              {
                DomainName: domainName,
              },
            ],
          },
          AliasTarget: {
            HostedZoneId: 'Z2FDTNDATAQYW2',
            DNSName: {
              'Fn::Sub': [
                '${DomainName}.',
                {
                  DomainName: {
                    'Fn::GetAtt': [`${id}CloudFront`, 'DomainName'],
                  },
                },
              ],
            },
          },
        },
      },
    },
  };
});
