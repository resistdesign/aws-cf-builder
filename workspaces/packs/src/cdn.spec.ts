import { addCDN } from './cdn';

describe('CDN', () => {
  describe('addCDN', () => {
    test('should add a CDN', () => {
      const cft = addCDN(
        {
          id: 'Basic',
          hostedZoneId: '1234567890',
          domainName: 'app.example.com',
          fileStorageId: 'BasicS3',
          certificateArn: 'i:am:a:cert:arn',
        },
        { AWSTemplateFormatVersion: '2010-09-09' }
      );

      expect(cft).toMatchSnapshot();
    });
  });
});
