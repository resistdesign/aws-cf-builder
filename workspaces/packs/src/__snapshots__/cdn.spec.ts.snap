// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDN addCDN should add a CDN 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Resources": Object {
    "BasicCloudFront": Object {
      "DependsOn": "BasicS3",
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "app.example.com",
          ],
          "CustomErrorResponses": Array [
            Object {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            Object {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "ForwardedValues": Object {
              "QueryString": false,
            },
            "TargetOriginId": Object {
              "Fn::Sub": Array [
                "S3-\${S3BucketName}",
                Object {
                  "S3BucketName": "app.example.com",
                },
              ],
            },
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": false,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::Sub": Array [
                  "\${S3BucketName}.s3.amazonaws.com",
                  Object {
                    "S3BucketName": "app.example.com",
                  },
                ],
              },
              "Id": Object {
                "Fn::Sub": Array [
                  "S3-\${S3BucketName}",
                  Object {
                    "S3BucketName": "app.example.com",
                  },
                ],
              },
              "S3OriginConfig": Object {
                "OriginAccessIdentity": "",
              },
            },
          ],
          "PriceClass": "PriceClass_All",
          "ViewerCertificate": Object {
            "AcmCertificateArn": "i:am:a:cert:arn",
            "MinimumProtocolVersion": "TLSv1.1_2016",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "BasicRoute53Record": Object {
      "DependsOn": Array [
        "BasicCloudFront",
      ],
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::Sub": Array [
              "\${DomainName}.",
              Object {
                "DomainName": Object {
                  "Fn::GetAtt": Array [
                    "BasicCloudFront",
                    "DomainName",
                  ],
                },
              },
            ],
          },
          "HostedZoneId": "Z2FDTNDATAQYW2",
        },
        "HostedZoneId": "1234567890",
        "Name": Object {
          "Fn::Sub": Array [
            "\${DomainName}.",
            Object {
              "DomainName": "app.example.com",
            },
          ],
        },
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
}
`;
